runtime/lib/backend/LightningKokkosSimulator.cpp

```diff
@@ -131,58 +131,105 @@ void LightningKokkosSimulator::NamedOperation(const std::string &name,
+    }
 }
 
 auto LightningKokkosSimulator::Observable(ObsId id, const std::vector<std::complex<double>> &matrix,
                                           const std::vector<QubitIdType> &wires) -> ObsIdType
 {
-    if (id == ObsId::Hermitian) {
-        throw std::logic_error(
-            "Hermitian observable not implemented in PennyLane-Lightning-Kokkos");
-    }
-
-    QFailIf(wires.size() > GetNumQubits(), "Invalid number of wires");
+    QFailIf(wires.size() > this->GetNumQubits(), "Invalid number of wires");
     QFailIf(!isValidQubits(wires), "Invalid given wires");
 
     auto &&dev_wires = getDeviceWires(wires);
+
+    if (id == ObsId::Hermitian) {
+        return this->obs_manager.createHermitianObs(matrix, dev_wires);
+    }
+
     return this->obs_manager.createNamedObs(id, dev_wires);
 }
 
 auto LightningKokkosSimulator::TensorObservable(const std::vector<ObsIdType> &obs) -> ObsIdType
 {
-    throw std::logic_error("Tensor observable not implemented in PennyLane-Lightning-Kokkos");
+    return this->obs_manager.createTensorProdObs(obs);
 }
 
 auto LightningKokkosSimulator::HamiltonianObservable(const std::vector<double> &coeffs,
                                                      const std::vector<ObsIdType> &obs) -> ObsIdType
 {
-    throw std::logic_error("Hamiltonian observable not implemented in PennyLane-Lightning-Kokkos");
+    return this->obs_manager.createHamiltonianObs(coeffs, obs);
 }
 
 auto LightningKokkosSimulator::Expval(ObsIdType obsKey) -> double
 {
+    using UnmanagedComplexHostView = Kokkos::View<Kokkos::complex<double> *, Kokkos::HostSpace,
+                                                  Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+
     QFailIf(!this->obs_manager.isValidObservables({obsKey}), "Invalid key for cached observables");
-    auto &&[obs, wires] = this->obs_manager.getObservable(obsKey);
 
     // update tape caching
     if (this->cache_recording) {
         cache_manager.addObservable(obsKey, Lightning::Measurements::Expval);
     }
 
-    if (obs == ObsId::PauliX) {
-        return this->device_sv->getExpectationValuePauliX(wires);
-    }
-    else if (obs == ObsId::PauliY) {
-        return this->device_sv->getExpectationValuePauliY(wires);
+    auto &&[obs, obs_type] = this->obs_manager.getObservable(obsKey);
+
+    if (obs_type != ObsType::Basic) {
+        std::cerr << "Warning: Expectation value of none-basic observables not supported yet in "
+                     "PennyLane-Lightning-Kokkos; the given observable is recorded for computing "
+                     "the runtime gradient algorithm."
+                  << std::endl;
+        return std::numeric_limits<double>::max();
     }
-    else if (obs == ObsId::PauliZ) {
-        return this->device_sv->getExpectationValuePauliZ(wires);
+
+    auto &&obs_name = obs->getObsName();
+    size_t pos = obs_name.find_first_of("[0123456789");
+    if (pos != std::string::npos) {
+        obs_name = obs_name.substr(0, pos);
     }
```

Comment:
What does the `find_first_of` do in this context?

Reply:
The `obs_name` is a concatenation of observable names and wires (and matrix). Lines 209-212 splits the observable name from the rest of the returned string.

Reply:
Are you saying the observable name has the entire matrix in it? 😅

Reply:
Not the entire matrix, but its hash would be part of the name.

Ah! Sorry, I just noticed I did make this confusion in my previous answer. Fixed it! :)

Reply:
Got it, thanks!